### PR TITLE
Replace python backend with lua

### PR DESCRIPTION
**Pull request by hkupty**
This makes the lua backend actually accessible through mappings and commands.

It also removes the python version.

If you want to test this before the actual merging of this PR:

```vim
Plug 'vigemus/iron.nvim', { 'branch': 'lua/replace' }
```

This PR will be merged - if everything is fine - on 08/20.

Please test this branch - every feedback is valuable.

### CODE REVIEW

**Comment by rzpilgrim** [comment]
Nice work! It's working with the remote kernel (#84). How can I change the open command? The following is in my `init.vim`:

```vimscript
set splitright
let g:iron_repl_open_cmd = 'vsplit'
```

So IronRepl should open a split to the right, but it is opening a (fixed-width?) split to the left.

**Comment by hkupty** [comment]
@rzpilgrim iron now is 100% lua, so even for configuring it you must do in lua.
You can do the following, on your `init.vim`

```vim
lua << EOF
local iron = require("iron")

iron.core.set_config{
  repl_open_cmd = "vsplit"
}
EOF
```

**Comment by hkupty** [comment]
As for the error, I'll check that.
For some reason it's treating the input as single string, not as a list of lines.

**Comment by hkupty** [comment]
@rzpilgrim Please check again. ea9eef5 should have fixed your issue.

**Comment by rzpilgrim** [comment]
Just tried again using ea9eef5. No dice (I got the same error as before).

**Comment by hkupty** [comment]
Weird. I thought I had solved it. Can you give more details?

File: lua/iron/fts/common.lua [hunk 1]
```diff
@@ -6,26 +6,26 @@ local fthelper = {
 fthelper.functions.format = function(repldef, lines)
   local tp = fthelper.types[repldef.type or "plain"](repldef)
   local new = {}
-  local off = 0
 
   if tp.open ~= nil then
-    new = {tp.open}
-    off = 1
+    new = table.insert(tp.open)
```
Comment (rzpilgrim) [inline lua/iron/fts/common.lua]: This appears to be the offending line:

```
Error detected while processing function IronSendMotion:
line    1:
E5105: Error while calling lua chunk: .../.config/nvim/plugged/iron.nvim//lua/iron/fts/common.lua:11: bad argument #1 to 'insert' (table expected, got string)
```

**Comment by rzpilgrim** [comment]
Like I said, same as before:

```
Error detected while processing function IronSendMotion:
line    1:
E5105: Error while calling lua chunk: .../.config/nvim/plugged/iron.nvim//lua/iron/fts/common.lua:11: bad argument #1 to 'insert' (table expected, got string)
```

I have tried sending both a visual selection and sending to the end of the line, and they both give the same error.

**Comment by hkupty** [comment]
@rzpilgrim Ok, so I was able to reproduce your issue and it was fixed with tests on the latest commit.
 Thanks for spotting that!

**Comment by rzpilgrim** [comment]
Thanks for the fix! I can now send text, but the behavior is not what it was before. If I visually select two lines, and the cursor is in the middle of the last line, then the character before the cursor will be the last sent to the REPL.

Before, sending a visual selection would send the entire highlighted region.

**Comment by hkupty** [comment]
I think this has to do with escape sequences.

I'll add integration tests so I can make sure we always send the correct data.

**Comment by hkupty** [comment]
So, as expected, since I still didn't fix the escape sequence issue, this PR won't be merged.

Please, if you could test more - as I suspect more issues would still exist - it'd be great before merging this.

New merge date is 1 week forward - 27/08

**Comment by hkupty** [comment]
@rzpilgrim Sorry, I delayed a bit but I think I managed to fix the issue.
One thing I couldn't test is the OSX to Linux through jupyter. Would you mind checking again?

I'll give some more room for this to be merged. Meanwhile I'll add some more tests and logs.
Cheers

**Comment by phillipm** [comment]
Hey, I just gave this branch a shot on a clojure file and when I run `:IronRepl` it now opens a normal empty buffer instead of a terminal buffer and doesn't actually start a clojure repl. I tried neovim 0.3.1 as well as the git build. Let me know if there are more details I can provide

**Comment by hkupty** [comment]
@phillipm Thanks for pointing that out. I completely missed that. I'll try to fix it this week.
Cheers mate!

**Comment by tobanw** [comment]
Great plugin, just found it and am trying out this branch. I found some issues with sending a chunk selected in "Visual Line" (`V`) mode.

1. If the selection extends downward from the cursor (e.g., `Vjctr`), it only sends the text from the point where the cursor was when `V` was pressed, not the entirety of the first line (as you would get from yanking the selection).
2. If the selection extends upward from the cursor (e.g., `Vkctr`), it throws an error:
```
Error detected while processing function IronSendMotion:                                                                       
line    4:                                                                                                                     
E5105: Error while calling lua chunk: ...n/.local/share/nvim/plugged/iron.nvim//lua/iron/init.lua:228: bad argument #1 to 'sub' (string expected, got nil) 
```

Otherwise, it works as expected in regular "Visual" mode.

**Comment by tobanw** [comment]
I found a quirk with Python: code chunks are always sent over with a leading and a trailing linebreak. This causes the standard `python` REPL to throw an error. In `ipython`, the code runs fine but you can see the linebreaks, e.g.,
```
In[1]:
   ...:
   ...: x = 1
   ...:
```

**Comment by phillipm** [comment]
@hkupty regarding my issue with clojure repls, it has to do with repl commands that have spaces in them. Like if I replace `lein repl` with `clj` it works.
I (blindly) changed `{{repl.command}}` to `{repl.command}` in the [init.lua](https://github.com/Vigemus/iron.nvim/blob/0bc1a1f1711a33e8c6c7ecb1d5292555c7a6b304/lua/iron/init.lua#L106) and it fixed the issue.

**Comment by hkupty** [comment]
That will work for clojure and break for the other repls, which require a
single command (such as `python` or `lua`).

I'll fix this with a correct 'list of args' approach, which should work for
all...

**Comment by hkupty** [comment]
@phillipm I've fixed the issue with lein/clojure repls.
With that said, I still have to check the windows split thing mentioned on #89.

After that, I'll merge this branch.

**Comment by hkupty** [comment]
All green. Merging!

**Comment by danjenson** [comment]
This still seems to be an issue; when sending a block selected in visual mode, it starts where the cursor switched into visual mode and copies thereafter. The cursor has to be in the first character of the first line in order for it to send properly. It also prefixes and postfixes and empty line, causing the ipython interpreter to balk when the spacing is not justified. python 3.7.2 and ipython 7.2.0 on Fedora 29

**Pull request merged**